### PR TITLE
add get account type code endpoint

### PR DESCRIPTION
**Pull request by sleepyzzzzzz**
## Changes made in this PR
Implement an endpoint for getting account type code by a given account type id

## Corresponding noted Ruby branch

def user_base_model
          sql_adapter[:user]
              .join(:accounts, id: :users__account_id)
              .join(:account_types, id: :accounts__type_id)
              .join(:user_access_tokens, user_id: :users__id)
              .left_join(:user_to_roles, user_id: :users__id)
              .left_join(:intercom_users, email: :users__email)
              .where { expires_at > Time.now }
              .select_all(:users)
              .select_append(:accounts__name___account_name)
              .select_append(:users__administrator___admin)
              .select_append(:accounts__parent_id, :accounts__receiver_id, :account_types__code___account_type)
              .select_append(:user_access_tokens__token)
              .select_append(Sequel.function(:group_concat, :user_to_roles__role_id).distinct.as(:user_role_ids))
              .select_append(:intercom_users__id_key___intercom_user_id_key)
              .group_by(:users__id)
              .order(Sequel.desc(:user_access_tokens__id))
        end

## TODO - changes that needs to be made
< leave the todos here to help the reviewers know what's the remaining pieces that haven't been implemented >
                                                                                        
## Implementation Checklist
Please check following check list to ensure the tasks are considered and completed in this branch.

#### API Design
- [x] Defined gRPC API in proto file
- [x] regenerated go code using `make proto_generate` command

#### API implementation
- [x] Implemented API in corresponding gRPC server
- [x] Added tests in gRPC layer

#### Domain Service and DB query Implementation
- [x] Added new query in dbstore layer
- [x] Added tests in db store layer
                                                                                       
- [x] Implemented functionality in Domain Service Layer
- [x] Added test in Domain Service Layer

- [ ] Created a new branch in paramount-api, with comments there indicating the functionality in Ruby that's re-implemented in Golang

#### Postgres Layer - identify and potentially add db migration to create index for fast querying.
- [ ] Verified and potentially created index for the newly added query.

### CODE REVIEW

File: apis/pdm/accounts/v1/accounts_service.proto [hunk 1]
```diff
@@ -38,6 +38,7 @@ service AccountsService {
   rpc GetPermissions(GetPermissionsRequest) returns (GetPermissionsResponse);
   rpc GetPermissionsByAccountID(GetPermissionsByAccountIDRequest) returns (GetPermissionsByAccountIDResponse);
   rpc HasPermission(HasPermissionRequest) returns (HasPermissionResponse);
+  rpc GetCodeByAccountTypeID(GetCodeByAccountTypeIDRequest) returns (GetCodeByAccountTypeIDResponse);
```
Comment (lhan-the-real) [inline apis/pdm/accounts/v1/accounts_service.proto:41]: I'm wondering if the account type code is only used as part of the account info. If so, what about we just add the code as part of the get account response? Instead of just returning the account type id, we return account type id and account type code.

Comment (lhan-the-real) [reply apis/pdm/accounts/v1/accounts_service.proto:41]: If this is not used or implemented, let's delete it.

File: apis/pdm/accounts/v1/accounts_service.proto [hunk 2]
```diff
@@ -46,6 +47,7 @@ message GetAccountRequest {
 
 message GetAccountResponse {
   Account account = 1;
+  string type_code = 2;
```
Comment (lhan-the-real) [inline apis/pdm/accounts/v1/accounts_service.proto:50]: I suggest we add the account_type_code into grpc message Account{}. So, for other account related apis, we are also returning the account type code as part of the response.

File: internal/domain/accounts/store.go [hunk 3]
```diff
@@ -744,3 +748,27 @@ func (a store) GetPermissionIDsByAccountType(
 
 	return permissionIDs, nil
 }
+
+func (a store) GetCodeByAccountType(
```
Comment (lhan-the-real) [inline internal/domain/accounts/store.go:752]: Let's add a comment to describe the code functionality.

File: internal/domain/accounts/domainService.go [hunk 4]
```diff
@@ -688,3 +688,10 @@ func (s *DomainService) GetIntegratedChannelAccountInfo(
 
 	return account, receiver, nil
 }
+
+func (s *DomainService) GetCodeByAccountType(
```
Comment (lhan-the-real) [inline internal/domain/accounts/domainService.go:692]: We didn't directly call the store function in other DomainService is because it's not part of the accounts.Account right?

File: internal/domain/accounts/store_test.go [hunk 5]
```diff
@@ -543,6 +543,26 @@ var _ = Describe("AccountStore", func() {
 				Expect(permissionIDs).To(Equal([]int32{}))
 			})
 		})
+
+		Describe("GetCodeByAccountType", func() {
+			It("should return an empty string for an invalid or nonexistent account type id", func() {
```
Comment (lhan-the-real) [inline internal/domain/accounts/store_test.go:548]: Add test cases with a none-exist account type id.

File: internal/domain/accounts/store_test.go [hunk 6]
```diff
@@ -543,6 +543,26 @@ var _ = Describe("AccountStore", func() {
...
+				code, err := accountsStore.GetCodeByAccountType(
+					context.Background(),
+					int32(-1),
+				)
+				Expect(err).To(BeNil())
+				Expect(code).To(Equal(""))
+			})
+
+			It("should return the account type code for the given accoun type id", func() {
```
Comment (lhan-the-real) [inline internal/domain/accounts/store_test.go:557]: Add test cases for 3 more account types (feel free to choose yourself)

File: apis/grpcservice/grpcaccounts/acountsGRPCServer.go [hunk 7]
```diff
@@ -340,8 +460,22 @@ func (s *accountsGRPCServer) ListAccountsWithPermissionsByID(
 		return nil, pdmerr.NewParent(err, "accountsGRPCServer: failed to list accounts from accounts domain service")
 	}
 
+	listedTypeCodes := []string{}
+	for _, accountPermission := range list {
```
Comment (lhan-the-real) [inline apis/grpcservice/grpcaccounts/acountsGRPCServer.go:464]: This means, we need to issue multiple requests to db to get each account type to code.
What about we get the account type info map in DB, and pick whatever we need in memory? Similar to the permissionsmap implementation.

Comment (sleepyzzzzzz) [reply apis/grpcservice/grpcaccounts/acountsGRPCServer.go:464]: I see. Will change

File: internal/domain/accounts/domainService.go (hunk 4 above)
Comment (sleepyzzzzzz) [reply internal/domain/accounts/domainService.go:692]: yes

File: internal/domain/accounts/store.go [hunk 8]
```diff
@@ -68,6 +68,7 @@ type Store interface {
 	GetPermissionIDsByAccountType(
 		ctx context.Context, typeID int32,
 	) ([]int32, error)
+	GetAccountTypeCodes(ctx context.Context) (accountTypeMap map[int32]AccountType, err error)
```
Comment (lhan-the-real) [inline internal/domain/accounts/store.go:71]: Nit: function name: GetAccountTypeMap

Comment (lhan-the-real) [reply internal/domain/accounts/store.go:71]: Is there any other information in account type we want to provide besides the Code?

Comment (sleepyzzzzzz) [reply internal/domain/accounts/store.go:71]: name

**Review by lhan-the-real — APPROVED** [review]
